src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,911 @@
+
+  void renew_enqueue_buffer_if_needed();
+
+  void task_stacktrace(JfrSampleType type, JavaThread** last_thread);
+  JfrCPUTimeThreadSampler(int64_t period_millis, double throttle, u4 max_traces, u4 max_frames_per_trace);
+  ~JfrCPUTimeThreadSampler();
+
+  void start_thread();
+
+  void enroll();
+  void disenroll();
+  void update_all_thread_timers();
+  void set_sampling_period(int64_t period_millis);
+  void set_throttle(double throttle);
+
+  void process_trace_queue();
+
+ protected:
+    virtual void post_run();
+   public:
+    virtual const char* name() const { return "JFR CPU Time Thread Sampler"; }
+    virtual const char* type_name() const { return "JfrCPUTimeThreadSampler"; }
+    bool is_JfrSampler_thread() const { return true; }
+    void run();
+    void on_javathread_create(JavaThread* thread);
+    bool create_timer_for_thread(JavaThread* thread, timer_t &timerid);
+    void set_timer_time(timer_t timerid, int64_t period_millis);
+    void on_javathread_terminate(JavaThread* thread);
+    int64_t get_sampling_period() const { return Atomic::load(&_period_millis); };
+    int64_t get_actual_sampling_period() const { return Atomic::load(&_actual_period_millis); };
+    bool recompute_actual_sampling_period();
+
+    void handle_timer_signal(void* context);
+    void init_timers();
+    void stop_timer();
+};
+
+
+JfrCPUTimeThreadSampler::JfrCPUTimeThreadSampler(int64_t period_millis, double throttle, u4 max_traces, u4 max_frames_per_trace) :
+  _sample(),
+  _sampler_thread(nullptr),
+  _queues(max_traces, max_frames_per_trace),
+  _period_millis(period_millis),
+  _throttle(throttle),
+  _max_frames_per_trace(max_frames_per_trace),
+  _disenrolled(true),
+  _jfrFrames(JfrCHeapObj::new_array<JfrStackFrame>(_max_frames_per_trace)) {
+  assert(_period_millis >= 0 || _period_millis == -1, "invariant");
+}
+
+JfrCPUTimeThreadSampler::~JfrCPUTimeThreadSampler() {
+  JfrCHeapObj::free(_jfrFrames, sizeof(JfrStackFrame) * _max_frames_per_trace);
+}
+
+void JfrCPUTimeThreadSampler::on_javathread_create(JavaThread* thread) {
+  if (thread->is_Compiler_thread()) {
+    return;
+  }
+  if (thread->jfr_thread_local() != nullptr) {
+    timer_t timerid;
+    if (create_timer_for_thread(thread, timerid) && thread->jfr_thread_local() != nullptr) {
+      thread->jfr_thread_local()->set_timerid(timerid);
+    }
+  }
+}
+
+void JfrCPUTimeThreadSampler::on_javathread_terminate(JavaThread* thread) {
+  if (thread->jfr_thread_local() != nullptr && thread->jfr_thread_local()->has_timerid()) {
+    timer_delete(thread->jfr_thread_local()->timerid());
+    thread->jfr_thread_local()->unset_timerid();
+  }
+}
+
+bool JfrCPUTimeThreadSampler::recompute_actual_sampling_period() {
+  int64_t period_millis = get_sampling_period();
+  double throttle = Atomic::load(&_throttle);
+  int64_t old_value = get_actual_sampling_period();
+  int64_t new_value = 0;
+  if (throttle < 0) { // off
+    new_value = period_millis;
```

Comment:
Here `new_value` can be set to `-1` thus causing an invalid interval passed to `timer_settime` later.

Reply:
Good catch, fixed it.